Use size_t in insert_clip

mstabrin opened the pull request:
Hello Steve,

The problem is related to the observation you made, but at another place.
I replaced the int types with size_t types and now it is also working with large EMData objects.

However, I am far away from being an expert in C++, so feedback is highly appreciated (e.g. I am not sure it if it is better to make all variables size_t or cast them as size_t if they are needed).

The actual problem are the pointers src and dst in line 1831 and 1832, which caused the memcpy function to throw the segmentation fault.

Best,
Markus

mstabrin commented:
Unfortunately, it looks like some tests fail now.
I will look into it tomorrow.

sludtke42 commented:
No, most of these changes don't make sense. We do not support nx, ny or nz > 2^31. The only variables it is appropriate to change to 64 bits are variables containing multidimensional products, or pointers/indices into arrays. That is, we do not support or intend to support 1-D EMData objects >8gb of RAM. If you have a use-case where this is required, we should discuss it. However, it is fine to have 2-D or 3-D objects where nx*ny*nz > 2^31. To be honest, the only case we ever seriously considered along those lines was 3-D volumes. The idea of needing even a 2-D array with more than 2 billion values seemed pretty unlikely even in recent years.   Can I ask what it is that you are trying to do with such large 2-D "images" ?

--------------------------------------------------------------------------------------
Steven Ludtke, Ph.D. <sludtke@bcm.edu <mailto:sludtke@bcm.edu>>                      Baylor College of Medicine 
Charles C. Bell Jr., Professor of Structural Biology
Dept. of Biochemistry and Molecular Biology                      (www.bcm.edu/biochem <http://www.bcm.edu/biochem>)
Academic Director, CryoEM Core                                        (cryoem.bcm.edu <http://cryoem.bcm.edu/>)
Co-Director CIBR Center                                    (www.bcm.edu/research/cibr <http://www.bcm.edu/research/cibr>)



> On Feb 20, 2019, at 7:24 PM, Markus Stabrin <notifications@github.com> wrote:
> 
> @mstabrin <https://github.com/mstabrin> requested your review on: #310 <https://github.com/cryoem/eman2/pull/310> Use size_t in insert_clip.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/cryoem/eman2/pull/310#event-2153661911>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AJ6i6nK4GdKvA-RAkYxcARkJmiBFoMJmks5vPfVdgaJpZM4bGdEM>.
>

mstabrin commented:
Hello Steve,
Thank you for your feedback :)
I now use type casting only during multiplication and added a test to cover the failing scenario.
However, it turns out that the server does not provide enough memory for the failing scenario, so I comment the test out for now.

The failing scenario happened during our refinement with a small delta and we are creating a stack within MPI shared memory containing the projections in a 1D polar coordinates format.

Best,
Markus

pawelpap9 commented:
Hi

There must be some kind of confusion here. I do not know what you are trying to do and why, but only very naive coding would result in arrays of this size.  You can check existing code in SPARX and and see that this is handled in a rational way. 

Regards,
Pawel

> On Feb 21, 2019, at 4:10 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> Hello Steve,
> Thank you for your feedback :)
> I now use type casting only during multiplication and added a test to cover the failing scenario.
> However, it turns out that the server does not provide enough memory for the failing scenario, so I comment the test out for now.
> 
> The failing scenario happened during our refinement with a small delta and we are creating a stack within MPI shared memory containing the projections in a 1D polar coordinates format.
> 
> Best,
> Markus
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

mstabrin commented:
Hello Pawel,

the line that causes the segmentation fault https://github.com/cryoem/eman2/blame/master/sparx/bin/sxmeridien.py#L5811
which is of course a result of the EMData object "bigbuffer"  being super large and is initialised around this line https://github.com/cryoem/eman2/blame/master/sparx/bin/sxmeridien.py#L5736.

Since you did the implementation, I am open for alternative design suggestions :)

The problem occurs, once the refinement changes to a delta of 0.2 degrees.

Best,
Markus

pawelpap9 commented:
It does not. I just ran it yesterday with this step and it runs perfectly fine. 

What would be the image size at this moment?

Regards,
Pawel

> On Feb 21, 2019, at 5:33 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> Hello Pawel,
> 
> the line that causes the segmentation fault https://github.com/cryoem/eman2/blame/master/sparx/bin/sxmeridien.py#L5811
> which is of course a result of the EMData object "bigbuffer" being super large and is initialised around this line https://github.com/cryoem/eman2/blame/master/sparx/bin/sxmeridien.py#L5736.
> 
> Since you did the implementation, I am open for alternative design suggestions :)
> 
> The problem occurs, once the refinement changes to a delta of 0.2 degrees.
> 
> Best,
> Markus
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

mstabrin commented:
I add the output of Meridien at this stage.
It might be due to a case of heavy preferred orientations, meaning only side views are present.

2019-02-20_13:58:41 => ITERATION  #37. Current state:     RESTRICTED, nxinit: 200, delta:    0.2344, xr:    1.3102, ts:    0.4367
2019-02-20_13:58:41 => Executed successfully:  mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
2019-02-20_13:58:41 => Executed successfully:  mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037/oldparamstructure
2019-02-20_14:36:36 => PROJECTION MATCHING parameters of buffered local polar
                    =>  acc_rot              :   0.594097243415
                    =>  acc_trans            :   0.582291709052
                    =>  an                   :   1.40625
                    =>  anger                :   1e+23
                    =>  avgvaradj            :   [76318.71875, 76334.265625]
                    =>  bestres              :   71
                    =>  bestres_143          :   85
                    =>  currentres           :   71
                    =>  delta                :   0.234375
                    =>  directory            :   02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
                    =>  fsc143               :   85
                    =>  keepfirst            :   -1
                    =>  large_at_Nyquist     :   False
                    =>  lentop               :   2000
                    =>  local                :   False
                    =>  mainiteration        :   37
                    =>  maxfrad              :   100
                    =>  nima_per_chunk       :   [157902, 162158]
                    =>  no_improvement       :   0
                    =>  no_params_changes    :   0
                    =>  nxinit               :   200
                    =>  nxpolar              :   200
                    =>  nxstep               :   29
                    =>  pixercutoff          :   2.0
                    =>  previousoutputdir    :   02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036
                    =>  refvol               :   02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036/vol_0_036.hdf
                    =>  shifter              :   1e+23
                    =>  state                :   RESTRICTED
                    =>  ts                   :   0.436718781789
                    =>  xr                   :   1.31015634537
 number_of_cones:  1

pawelpap9 commented:
Distribution of data does not affect reference projections, at least not like
that. 

Just few lines above you can see the program estimates how many reference projections you can compute based on memory provided by the user. What this was set to?

An obvious thing is to do is to use smaller number in input. 

Regards,
Pawel

> On Feb 21, 2019, at 5:38 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> I add the output of Meridien at this stage.
> It might be due to a case of heavy preferred orientations, meaning only side views are present.
> 
> 2019-02-20_13:58:41 => ITERATION #37. Current state: RESTRICTED, nxinit: 200, delta: 0.2344, xr: 1.3102, ts: 0.4367
> 2019-02-20_13:58:41 => Executed successfully: mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
> 2019-02-20_13:58:41 => Executed successfully: mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037/oldparamstructure
> 2019-02-20_14:36:36 => PROJECTION MATCHING parameters of buffered local polar
> => acc_rot : 0.594097243415
> => acc_trans : 0.582291709052
> => an : 1.40625
> => anger : 1e+23
> => avgvaradj : [76318.71875, 76334.265625]
> => bestres : 71
> => bestres_143 : 85
> => currentres : 71
> => delta : 0.234375
> => directory : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
> => fsc143 : 85
> => keepfirst : -1
> => large_at_Nyquist : False
> => lentop : 2000
> => local : False
> => mainiteration : 37
> => maxfrad : 100
> => nima_per_chunk : [157902, 162158]
> => no_improvement : 0
> => no_params_changes : 0
> => nxinit : 200
> => nxpolar : 200
> => nxstep : 29
> => pixercutoff : 2.0
> => previousoutputdir : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036
> => refvol : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036/vol_0_036.hdf
> => shifter : 1e+23
> => state : RESTRICTED
> => ts : 0.436718781789
> => xr : 1.31015634537
> number_of_cones: 1
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

pawelpap9 commented:
You can uncomment line 5598 to see what program estimated. 

Regards,
Pawel

> On Feb 21, 2019, at 5:38 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> I add the output of Meridien at this stage.
> It might be due to a case of heavy preferred orientations, meaning only side views are present.
> 
> 2019-02-20_13:58:41 => ITERATION #37. Current state: RESTRICTED, nxinit: 200, delta: 0.2344, xr: 1.3102, ts: 0.4367
> 2019-02-20_13:58:41 => Executed successfully: mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
> 2019-02-20_13:58:41 => Executed successfully: mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037/oldparamstructure
> 2019-02-20_14:36:36 => PROJECTION MATCHING parameters of buffered local polar
> => acc_rot : 0.594097243415
> => acc_trans : 0.582291709052
> => an : 1.40625
> => anger : 1e+23
> => avgvaradj : [76318.71875, 76334.265625]
> => bestres : 71
> => bestres_143 : 85
> => currentres : 71
> => delta : 0.234375
> => directory : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
> => fsc143 : 85
> => keepfirst : -1
> => large_at_Nyquist : False
> => lentop : 2000
> => local : False
> => mainiteration : 37
> => maxfrad : 100
> => nima_per_chunk : [157902, 162158]
> => no_improvement : 0
> => no_params_changes : 0
> => nxinit : 200
> => nxpolar : 200
> => nxstep : 29
> => pixercutoff : 2.0
> => previousoutputdir : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036
> => refvol : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036/vol_0_036.hdf
> => shifter : 1e+23
> => state : RESTRICTED
> => ts : 0.436718781789
> => xr : 1.31015634537
> number_of_cones: 1
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

pawelpap9 commented:
Hi,

I am still puzzled where you got yesterday's dimensions from.

The reference images (which are 1D and relatively short, few thousand elements)  are inserted into bigbuffer,
which is created as 2D:

bigbuffer.insert_clip(crefim,(0,i) )

The problem may happen if references are small, as in your case, which means program will decide it can
use very large number of references and index i becomes very large.

In other words, dimensions of of bigbuffer are (small,  excessively large).

Steve mentioned yesterday that there is a limit on a size of an array dimension in EMAN2, but I do not think
the system was coded in a consistent manner this way, if it were, there would be a check when EMData objects
are created.

A simple possibility is to replace insert_clip by very simple functions I wrote and fixed quite recently to work
with large arrays, which are Util.get_slice and Util.put_slice.  The code is very simple (in util_sparx.cpp)
so you can easily make sure that indexes are of correct type/size.

____________________
Pawel

> On Feb 21, 2019, at 5:38 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> I add the output of Meridien at this stage.
> It might be due to a case of heavy preferred orientations, meaning only side views are present.
> 
> 2019-02-20_13:58:41 => ITERATION #37. Current state: RESTRICTED, nxinit: 200, delta: 0.2344, xr: 1.3102, ts: 0.4367
> 2019-02-20_13:58:41 => Executed successfully: mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
> 2019-02-20_13:58:41 => Executed successfully: mkdir 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037/oldparamstructure
> 2019-02-20_14:36:36 => PROJECTION MATCHING parameters of buffered local polar
> => acc_rot : 0.594097243415
> => acc_trans : 0.582291709052
> => an : 1.40625
> => anger : 1e+23
> => avgvaradj : [76318.71875, 76334.265625]
> => bestres : 71
> => bestres_143 : 85
> => currentres : 71
> => delta : 0.234375
> => directory : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main037
> => fsc143 : 85
> => keepfirst : -1
> => large_at_Nyquist : False
> => lentop : 2000
> => local : False
> => mainiteration : 37
> => maxfrad : 100
> => nima_per_chunk : [157902, 162158]
> => no_improvement : 0
> => no_params_changes : 0
> => nxinit : 200
> => nxpolar : 200
> => nxstep : 29
> => pixercutoff : 2.0
> => previousoutputdir : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036
> => refvol : 02_Meridien_SPA/01_Meridien_SPA_fromScratch_lp25_7.5degree_skipPreAli/main036/vol_0_036.hdf
> => shifter : 1e+23
> => state : RESTRICTED
> => ts : 0.436718781789
> => xr : 1.31015634537
> number_of_cones: 1
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mstabrin commented:
Hello Pawel,

Now I also think that we have some confusion here and that we do not talk about the same thing :)

First, we do not have a memory problem at all (If one has more than 8Gb ^^).  
I just mentioned a memory problem related to the automatic test server that runs every time one pushes something to github as I added a unit test for this scenario.  

The problem was indeed the datatype of the variables as Steve mentioned, but not inside the EMData object but only in the insert_clip function (for now).
The array dimensions of the EMData object are also fine, the only problem was that insert_clip could not handle large pointers that are used to copy the memory from one object to another one.
That is why I adjusted the datatype inside of the length calculation, so we should not have any problems anymore :)

Best,
Markus

mstabrin commented:
@sludtke42 Is it ok like this? Do you plan to test it first?

Best,
Markus

sludtke42 commented:
I don't see any problems with the change as it is now. Just make sure the test block remains commented out. We don't want routine tests which allocate 8+ gb of RAM.
--------------------------------------------------------------------------------------
Steven Ludtke, Ph.D. <sludtke@bcm.edu>                      Baylor College of Medicine 
Charles C. Bell Jr., Professor of Structural Biology
Dept. of Biochemistry and Molecular Biology                      (www.bcm.edu/biochem)
Academic Director, CryoEM Core                                        (cryoem.bcm.edu)
Co-Director CIBR Center                                    (www.bcm.edu/research/cibr)



> On Feb 21, 2019, at 10:01 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> @sludtke42 <https://github.com/sludtke42> Is it ok like this? Do you plan to test it first?
> 
> Best,
> Markus
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cryoem/eman2/pull/310#issuecomment-466055611>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AJ6i6qJIo8AWZIzbG_fkmkAipg2ZF53Oks5vPsLXgaJpZM4bGdEM>.
>

pawelpap9 commented:
I meant the same thing. 

What I said is that if the program thinks there is less memory it will create smaller buffer. So if you want to just get it to run quickly, this is the easy way out. 

As for clip, I would advise you to use Util.put_slice, which are much simpler and even if you break something it will not affect the entire system. 

Regards,
Pawel

> On Feb 21, 2019, at 7:05 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> Hello Pawel,
> 
> Now I also think that we have some confusion here and that we do not talk about the same thing :)
> 
> First, we do not have a memory problem at all (If one has more than 8Gb ^^).
> I just mentioned a memory problem related to the automatic test server that runs every time one pushes something to github as I added a unit test for this scenario.
> 
> The problem was indeed the datatype of the variables as Steve mentioned, but not inside the EMData object but only in the insert_clip function (for now).
> The array dimensions of the EMData object are also fine, the only problem was that insert_clip could not handle large pointers that are used to copy the memory from one object to another one.
> That is why I adjusted the datatype inside of the length calculation, so we should not have any problems anymore :)
> 
> Best,
> Markus
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

pawelpap9 commented:
This will not get you far as further down the line buffer is used and you will likely run into further problems. 

You have to check all uses of buffer and consequently do them through put_slice and get_slice. 

Regards,
Pawel

> On Feb 21, 2019, at 10:12 AM, Markus Stabrin <notifications@github.com> wrote:
> 
> Merged #310 into master.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

pawelpap9 commented:
The problem is not with memory size but with indexes. One would have to go through fundamental functions and make sure that whenever they deal with indexing the variables should be of size_t type. 

In older code I did not pay attention to it but in newer C code I try to do it consistently. 

Regards,
Pawel

> On Feb 21, 2019, at 10:07 AM, Steve Ludtke <notifications@github.com> wrote:
> 
> I don't see any problems with the change as it is now. Just make sure the test block remains commented out. We don't want routine tests which allocate 8+ gb of RAM.
> --------------------------------------------------------------------------------------
> Steven Ludtke, Ph.D. <sludtke@bcm.edu> Baylor College of Medicine 
> Charles C. Bell Jr., Professor of Structural Biology
> Dept. of Biochemistry and Molecular Biology (www.bcm.edu/biochem)
> Academic Director, CryoEM Core (cryoem.bcm.edu)
> Co-Director CIBR Center (www.bcm.edu/research/cibr)
> 
> 
> 
> > On Feb 21, 2019, at 10:01 AM, Markus Stabrin <notifications@github.com> wrote:
> > 
> > @sludtke42 <https://github.com/sludtke42> Is it ok like this? Do you plan to test it first?
> > 
> > Best,
> > Markus
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub <https://github.com/cryoem/eman2/pull/310#issuecomment-466055611>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AJ6i6qJIo8AWZIzbG_fkmkAipg2ZF53Oks5vPsLXgaJpZM4bGdEM>.
> > 
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

sludtke42 commented:
The code he changed today is, at worst, harmless. I was not stating that it was going to fix every possible indexing problem in the system. Again, I do not plan on changing the single axis index variables from 32 to 64 bit anywhere.

When this issue first arose was with tomograms which had more than 2^31 voxels. At that time the system was updated so products of axis indices would use 64 bit ints, but the axis indices themselves remained 32 bits. This was done fairly consistently throughout the system, but nobody thought specifically of the possibility that nx*ny would require 64 bits, it was only nx*ny*nz which was considered in depth. So there probably still are a few bugs floating around the system related to this, and it isn't a bad idea to fix them when we see them...
 
--------------------------------------------------------------------------------------
Steven Ludtke, Ph.D. <sludtke@bcm.edu>                      Baylor College of Medicine 
Charles C. Bell Jr., Professor of Structural Biology
Dept. of Biochemistry and Molecular Biology                      (www.bcm.edu/biochem)
Academic Director, CryoEM Core                                        (cryoem.bcm.edu)
Co-Director CIBR Center                                    (www.bcm.edu/research/cibr)



> On Feb 21, 2019, at 10:39 AM, Pawel <notifications@github.com> wrote:
> 
> The problem is not with memory size but with indexes. One would have to go through fundamental functions and make sure that whenever they deal with indexing the variables should be of size_t type. 
> 
> In older code I did not pay attention to it but in newer C code I try to do it consistently. 
> 
> Regards,
> Pawel
> 
> > On Feb 21, 2019, at 10:07 AM, Steve Ludtke <notifications@github.com> wrote:
> > 
> > I don't see any problems with the change as it is now. Just make sure the test block remains commented out. We don't want routine tests which allocate 8+ gb of RAM.
> > --------------------------------------------------------------------------------------
> > Steven Ludtke, Ph.D. <sludtke@bcm.edu> Baylor College of Medicine 
> > Charles C. Bell Jr., Professor of Structural Biology
> > Dept. of Biochemistry and Molecular Biology (www.bcm.edu/biochem)
> > Academic Director, CryoEM Core (cryoem.bcm.edu)
> > Co-Director CIBR Center (www.bcm.edu/research/cibr)
> > 
> > 
> > 
> > > On Feb 21, 2019, at 10:01 AM, Markus Stabrin <notifications@github.com> wrote:
> > > 
> > > @sludtke42 <https://github.com/sludtke42> Is it ok like this? Do you plan to test it first?
> > > 
> > > Best,
> > > Markus
> > > 
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub <https://github.com/cryoem/eman2/pull/310#issuecomment-466055611>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AJ6i6qJIo8AWZIzbG_fkmkAipg2ZF53Oks5vPsLXgaJpZM4bGdEM>.
> > > 
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cryoem/eman2/pull/310#issuecomment-466070933>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AJ6i6n0GdM7nrMYdrJwlIOUjN4Ma5cbiks5vPsvMgaJpZM4bGdEM>.
>

pawelpap9 commented:
Hi,

not to beat a dead horse, but just as a way of explanation, as the issue is related to our discussion on Monday.

The bigbuffer is just a buffer to hold floats, not an image.  It originates as numpy, and it should, as it is generate
by mpi mechanism.  After writing the beginning I realized I have to turn into EMData as the real use of bigbuffer
is in various C functions, and as we discussed we do not have (or discourage) passing numpy to C level.
So, in reality this EMData object is just a pointer to numpy data.

In principle, there is nothing wrong with EMData holding non-image information, but it is somewhat awkward.
However, once the issue of indexes is solved, it should be fine.

Markus: I remember when I first tested the code I noticed that the program would crash if bigbuffer was too large.
However, the problem went away once I started using it within meridien.  Now I see that most likely the available
emory I have on my cluster is much less than what you have, so I am sure I never reach the breaking point.


____________________
Pawel
